pytorch/2. Autograd.ipynb

```diff
@@ -0,0 +1,988 @@
+   "execution_count": 196,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Graph : x + y\n",
+      "o.requires_grad : False \n",
+      "o.grad_fn : None \n",
+      "\n",
+      "\n",
+      "Graph : x + y + z\n",
+      "o.requires_grad : False \n",
+      "o.grad_fn : None \n"
+     ]
+    }
+   ],
+   "source": [
+    "print('Graph : x + y')\n",
+    "o = x + y\n",
+    "print('o.requires_grad : %s ' % (o.requires_grad))\n",
+    "print('o.grad_fn : %s ' % (o.grad_fn))\n",
+    "print('\\n\\nGraph : x + y + z')\n",
+    "o = x + y + z\n",
+    "print('o.requires_grad : %s ' % (o.requires_grad))\n",
+    "print('o.grad_fn : %s ' % (o.grad_fn))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### `.detach()`\n",
+    "\n",
+    "1. It is possible to detach variables from the graph by calling `.deatch()`. \n",
+    "2. This could lead to disconnected graphs. In which case PyTorch will only backpropagate gradients until the point of disconnection."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 41,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "x = Variable(torch.Tensor(3, 5).uniform_(-1, 1), requires_grad=True)\n",
+    "y = Variable(torch.Tensor(3, 5).uniform_(-1, 1), requires_grad=True)\n",
+    "z = Variable(torch.Tensor(3, 5).uniform_(-1, 1), requires_grad=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![caption](images/detach.png)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 42,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "m1 = x + y\n",
+    "m2 = z ** 2\n",
+    "m1 = m1.detach()\n",
+    "m3 = m1 + m2\n",
+    "m3.backward(torch.ones(3, 5))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 43,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "dm3/dx \n",
+      "\n",
+      " None \n",
+      "\n",
+      "dm3/dy \n",
+      "\n",
+      " None \n",
+      "\n",
+      "dm3/dz \n",
+      "\n",
+      " Variable containing:\n",
+      " 1.2043  1.9914  0.1340 -1.8074  1.3064\n",
+      "-0.1923  0.9834 -1.9299  1.4948  0.6174\n",
+      " 1.0566 -1.1677 -1.5411  0.5598 -1.9467\n",
+      "[torch.FloatTensor of size 3x5]\n",
+      " \n"
+     ]
+    }
+   ],
+   "source": [
+    "print('dm3/dx \\n\\n %s ' % (x.grad))\n",
+    "print('\\ndm3/dy \\n\\n %s ' % (y.grad))\n",
+    "print('\\ndm3/dz \\n\\n %s ' % (z.grad))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "## Gradients w.r.t intermediate variables in the graph\n",
+    "\n",
+    "1. By default, PyTorch all gradient computations w.r.t intermediate nodes in the graph are ad-hoc.\n",
```

Comment:
I'm not sure what that means

Reply:
Ah what I meant is that PyTorch does not retain the gradients w.r.t intermediate nodes in the graph in order to save memory. It only retains the gradient of the loss w.r.t to the things that we actually care about (like the parameters of our model) and the intermediate gradients are "ad-hoc" since they are only used to compute gradient of the loss w.r.t the model parameters. I'll make this clearer :)